### PR TITLE
Rotpup

### PR DESCRIPTION
**Pull request by jason-hobbs**
This now supports Rails 6.1, and also includes the latest rotp gem with syntax changes so the qr code works and it will correctly verify the entered code.

### CODE REVIEW

File: lib/devise_two_factor/models/two_factor_authenticatable.rb [hunk 1]
```diff
@@ -32,7 +32,8 @@ def validate_and_consume_otp!(code, options = {})
         return false unless code.present? && otp_secret.present?
 
         totp = self.otp(otp_secret)
-        return consume_otp! if totp.verify_with_drift(code, self.class.otp_allowed_drift)
+        # return consume_otp! if totp.verify_with_drift(code, self.class.otp_allowed_drift)
```
Comment (vimutter) [inline lib/devise_two_factor/models/two_factor_authenticatable.rb:35]: Hi, let's remove commented code.

**Review by vimutter — CHANGES_REQUESTED** [review]

File: lib/devise_two_factor/spec_helpers/two_factor_authenticatable_shared_examples.rb [hunk 2]
```diff
@@ -77,17 +77,17 @@
...
-      otp = ROTP::TOTP.new(otp_secret).at(Time.now + subject.class.otp_allowed_drift * 2, true)
+      otp = ROTP::TOTP.new(otp_secret).at(Time.now + subject.class.otp_allowed_drift * 2)
       expect(subject.validate_and_consume_otp!(otp)).to be false
     end
 
     it 'does not validate an OTP below the allowed drift' do
-      otp = ROTP::TOTP.new(otp_secret).at(Time.now - subject.class.otp_allowed_drift * 2, true)
+      otp = ROTP::TOTP.new(otp_secret).at(Time.now - subject.class.otp_allowed_drift * 2)
       expect(subject.validate_and_consume_otp!(otp)).to be false
     end
```
Comment (vimutter) [inline lib/devise_two_factor/spec_helpers/two_factor_authenticatable_shared_examples.rb:92]: Hi, line 95-100 we need to fix broken shared example too, now it has escaped email in generated code. So pattern checked needs to include `CGI.escape` of the account (fake email).

File: lib/devise_two_factor/models/two_factor_authenticatable.rb [hunk 3]
```diff
@@ -32,7 +32,8 @@ def validate_and_consume_otp!(code, options = {})
...
+        return consume_otp! if totp.verify(code, drift_behind: 30)
```
Comment (vimutter) [inline lib/devise_two_factor/models/two_factor_authenticatable.rb:36]:
```
if totp.verify(code, drift_behind: self.class.otp_allowed_drift, drift_ahead: self.class.otp_allowed_drift)
  return consume_otp! 
end
```

File: lib/devise_two_factor/spec_helpers/two_factor_authenticatable_shared_examples.rb (hunk 2 above)
Comment (vimutter) [reply lib/devise_two_factor/spec_helpers/two_factor_authenticatable_shared_examples.rb:92]:
```
describe '#otp_provisioning_uri' do
    let(:otp_secret_length) { subject.class.otp_secret_length }
    let(:account)           { Faker::Internet.email }
    let(:issuer)            { "Tinfoil" }

    it "should return uri with specified account" do
      expect(subject.otp_provisioning_uri(account)).to match(%r{otpauth://totp/#{CGI.escape account}\?secret=\w{#{otp_secret_length}}})
    end

    it 'should return uri with issuer option', :aggregate_failures do
      expect(subject.otp_provisioning_uri(account, issuer: issuer)).
        to match(%r{otpauth://totp/#{CGI.escape issuer}:#{CGI.escape account}\?.*secret=\w{#{otp_secret_length}}(&|$)})
    end
  end
```

**Review by vimutter — CHANGES_REQUESTED** [review]

**Comment by jason-hobbs** [comment]
Okay, changes have been added.

**Comment by igor-drozdov** [comment]
@vimutter @bsedat could you please have another look at this PR?

File: lib/devise_two_factor/models/two_factor_authenticatable.rb (hunk 3 above)
Comment (bsedat) [reply lib/devise_two_factor/models/two_factor_authenticatable.rb:35]: Would we be able to use the `otp_allowed_drift` for the drift tolerances?

**Comment by jonathansimmons** [comment]
@bsedat Now that you're back in maintenance mode eyes on this PR would be great. We're going to fork and patch in a project for now. 

Cc: @blakeprudhomme

**Comment by ebeeson** [comment]
This will be merged soon and then we'll cut a new release.

**Comment by ebeeson** [comment]
I believe everything from this has been included in v4.0.0.

**Pull request closed without merging**